temporal-sdk/src/main/java/io/temporal/internal/worker/Poller.java

```diff
@@ -232,27 +237,44 @@ public void run() {
           suspender.await();
         }
 
-        if (pollExecutor.isShutdown()) {
-          return;
-        }
+        interruptIfNeeded();
+
         task.run();
         pollBackoffThrottler.success();
+      } catch (InterruptedException e) {
+        // we restore the flag here, so it can be checked and processed (with exit) in finally
+        Thread.currentThread().interrupt();
       } catch (Throwable e) {
         pollBackoffThrottler.failure();
-        if (!pollExecutor.isShutdown() && !pollExecutor.isTerminating()
-            || !(e.getCause() instanceof InterruptedException)
-                && !(e instanceof RejectedExecutionException)) {
+        // if we are terminating and getting rejected execution - it's normal
+        if (!pollExecutor.isTerminating() || !(e instanceof RejectedExecutionException)) {
```

Comment:
`pollExecutor.isTerminating()` is covering both `pollExecutor.isShutdown()` and `pollExecutor.isTerminating()`.
`|| !(e.getCause() instanceof InterruptedException)` was here for handling what `LocalActtivityPollTask` was doing here:
```
catch (InterruptedException e) {
      throw new RuntimeException("local activity poll task interrupted", e);
}
```
It's now replaced with a proper interrupted flag handling.